Feature comments

laserdi opened the pull request.

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• initial commit.

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• прописана логика.
• тесты постмана ещё не написаны.

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• прописана логика.
• тесты постмана ещё не написаны.

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• прописана логика.
• исправлена конвертер комментариев.
• тесты постмана ещё не написаны.

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• прописана логика.
• добавлен энпоинт получения списка комментариев к событию.
• тесты постмана ещё не написаны.

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• обновил Readme.md и контроллер комментариев.

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• написаны тесты.

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• удалены тесты пройденных этапов.

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• изменён файл с тестами.

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• добавлена ссылка в readme.md

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/controller/CommentAdminController.java:
```diff
@@ -0,0 +1,32 @@
+     */
+    @DeleteMapping("/{comId}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void deleteByAdmin(@PathVariable @NotNull @PositiveOrZero Long comId) {
```
Действительно ли здесь требуется аннотация `@NotNull`? Параметр пути является обязательным, так как в `@PathVariable` не указано обратное - `defaultValue = ".."` или `required = false`

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/controller/CommentUserController.java:
```diff
@@ -0,0 +1,88 @@
+    @GetMapping("/{comId}/user/{userId}")
+    @ResponseStatus(HttpStatus.OK)
+    public CommentForView getByIdForUser(@PathVariable("userId") @Positive @NotNull Long userId,
+                                         @PathVariable("comId") @Positive @NotNull Long commId) {
```
Действительно ли здесь требуется аннотация `@NotNull`? Параметр пути является обязательным, так как в `@PathVariable` не указано обратное - `defaultValue = ".."` или `required = false`

laserdi replied:
Убрал.

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/controller/CommentUserController.java:
```diff
@@ -0,0 +1,88 @@
+
+    @GetMapping("/{comId}/user/{userId}")
+    @ResponseStatus(HttpStatus.OK)
+    public CommentForView getByIdForUser(@PathVariable("userId") @Positive @NotNull Long userId,
```
Действительно ли здесь требуется аннотация `@NotNull`? Параметр пути является обязательным, так как в `@PathVariable` не указано обратное - `defaultValue = ".."` или `required = false`

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/dto/comment/CommentUserDto.java:
```diff
@@ -0,0 +1,30 @@
+@NoArgsConstructor
+public class CommentUserDto {
+
+    @PositiveOrZero(groups = CreateObject.class, message = "Для создания комментария требуется передать ID события.")
```
Также следует использовать `@NotNull`, так как аннотация `@PositiveOrZero` допускает null)

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/controller/CommentUserController.java:
```diff
@@ -0,0 +1,88 @@
+     */
+    @GetMapping("/event/{eventId}")
+    @ResponseStatus(HttpStatus.OK)
+    public List<CommentForView> getAllCommentEvent(@PathVariable @PositiveOrZero @NotNull Long eventId,
```
Хорошо было бы также использовать необязательный параметр/заголовок, который бы использовался для аутентификации пользователя) То есть, если это значение не null - значит можно проверить то, что, если это владелец события, то независимо от того, публичное или нет, получить комментарии к событию)

laserdi replied:
Исправил.

TyutterinYakov replied:
Здесь осталась лишняя аннотация `@NotNull`

laserdi replied:
Убрал.

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,213 @@
+                "При создании комментария пользователем в базе данных не найдено событие с ID = %d.");
+        Comment comment = commentMapper.mapToModelFromDto(inputCommentDto, eventFromDb, userFromDb);
+        comment.setCreatedOn(LocalDateTime.now());
+        System.out.println("Печать комментария: \t" + comment);
```
Для логгирования мы теперь используем логгер)

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,213 @@
+    public CommentForView addComment(Long userId, CommentUserDto inputCommentDto) {
+        User userFromDb = userService.getUserOrThrow(userId, "При создании комментария пользователем " +
+                "в базе данных не найден пользователь с ID = %d.");
+        Event eventFromDb = eventService.getEventOrThrow(inputCommentDto.getEventId(),
```
Оставить комментарий можно только для опубликованного события) Исключение - для организатора события

laserdi replied:
Сделал.

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,213 @@
+    public CommentForView getCommentById(Long userId, Long comId) {
+        User userFromDb = userService.getUserOrThrow(userId, "При получении комментария пользователем " +
+                "в базе данных не найден пользователь с ID = %d.");
+        Comment commentFromDb = getCommentOrThrow(comId, "При получении комментария пользователем " +
```
Комментарий можно получить только у опубликованного события) Исключение - для владельца события

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,213 @@
+    public void deleteCommentByAdmin(Long comId) {
+        commentRepository.findById(comId).orElseThrow(() -> new NotFoundRecordInBD(String.format(
+                "При удалении комментария с ID = %d администратором комментарий не найден в БД.", comId)));
+        System.out.println("Найден комментарий перед удалением.");
```
Для логгирования мы теперь используем логгер)

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,213 @@
+     * @throws WrongFieldValueException не передан текст для обновления.
+     */
+    private Comment updateFieldsByUser(CommentUserDto newComment, Comment oldComment) {
+        if (newComment.getText() == null || newComment.getText().isBlank()) {
```
Действительно ли эта проверка требуется? В DTO `CommentUserDto` используется аннотация `@NotBlank` над этим полем

laserdi replied:
Эту аннотацию isBlanlk() забыл убрать.

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,213 @@
+     */
+    @Override
+    @Transactional
+    public CommentForView updateComment(Long comId, Long userId, CommentUserDto inputCommentDto) {
```
Хорошо было бы ввести ограничение на редактирование) Например, 1 час после публикации комментария

laserdi replied:
Сделал 2 часа.

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,213 @@
+    public List<CommentForView> getCommentsForEvent(Long eventId, int from, int size) {
+        Pageable pageable = PageRequest.of(
+                from == 0 ? 0 : (from / size), size);
+        List<Comment> comments = commentRepository.findAllByEvent_Id(eventId, pageable);
```
Событие должно быть опубликовано. Исключение - для владельца события

laserdi replied:
Сделал.

TyutterinYakov reviewed (commented):
Добрый день! Также хорошо было бы в ShortDto и FullDto для событий присоединять количество всех комментариев, чтобы пользователь, просматривая список событий, мог увидеть, где, например, другие пользователи более активны) Хорошего дня! Если возникнут вопросы - пиши в пачку, буду рад помочь🌅

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• исправления после первого ревью.

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• исправления после первого ревью.

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• исправления после первого ревью.
• исправил метод добавления комментария.

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• исправления после первого ревью.
• исправил тесты feature.json.

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• исправления после первого ревью.
• исправил тесты feature.json: был не верный параметр в в тесте на создание комментария №2.

laserdi commented:
Добрый день! Также хорошо было бы в ShortDto и FullDto для событий присоединять количество всех комментариев, чтобы пользователь, просматривая список событий, мог увидеть, где, например, другие пользователи более активны) Хорошего дня! Если возникнут вопросы - пиши в пачку, буду рад помочь🌅

> Добрый день! Также хорошо было бы в ShortDto и FullDto для событий присоединять количество всех комментариев, чтобы пользователь, просматривая список событий, мог увидеть, где, например, другие пользователи более активны) Хорошего дня! Если возникнут вопросы - пиши в пачку, буду рад помочь🌅

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/repository/CommentRepository.java:
```diff
@@ -0,0 +1,18 @@
+    List<Comment> getByEvent_Id(Long id, Pageable pageable);
+
+    List<Comment> findAllByEvent_Id(Long id, Pageable pageable);
+
```
Также можно было бы добавить метод вида:
`Map<Long, Integer> getCommentsEvents(List<Long> eventIds);`
Запрос на JPQL будет иметь подобный вид:
`select com.event.id, count(com) from Comment com where com.event.id = ?1 group by com.event.id`
Этот метод мы будем использовать, когда потребуется получить количество комментариев для коллекции событий)

laserdi replied:
Делал на скорую руку.
Спасибо за совет, про это понимаю.
Буду рефакторить после сдачи. Не думал об этом функционале.

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/service/event/EventServiceImpl.java:
```diff
@@ -66,8 +68,10 @@ public List<EventShortDto> getMyEvents(Long userId, Integer from, Integer size)
         Pageable pageable = PageRequest.of(from, size, Sort.by("id").ascending());
 
```
Тут нам потребуется метод `getCommentsEvents` из репозитория комментариев, что я предложил, чтобы не плодить n-лишние запросы)

laserdi replied:
Да, я понимаю, что, чтобы не было много запросов, надо сразу отправлять один запрос, отправляя список ID.

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,277 @@
+                "При создании комментария пользователем в базе данных не найдено событие с ID = %d.");
+        //Проверяем опубликовано ли событие
+        if (!eventFromDb.getEventState().equals(EventState.PUBLISHED)) {
+            if (eventFromDb.getInitiator().getId().equals(userId)) {
```
Можно было подправить условие таким образом:
```
if (!eventFromDb.getInitiator().getId().equals(userId)) {
    log.info("Пользователь пытается сделать комментарий неопубликованного события.");
    throw new OperationFailedException("Пользователь пытается сделать комментарий " +
                            "неопубликованного события.");
}
```
Чтобы не дублировать код) Так как после условий идет такая же логика по сохранению комментария)

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,277 @@
+        //Проверяем опубликовано ли событие
+        if (!eventFromDb.getEventState().equals(EventState.PUBLISHED)) {
+            //Событие не опубликовано.
+            if (eventFromDb.getInitiator().getId().equals(userId)) {
```
Аналогично, здесь также можно было бы изменить условие:
`if (!eventFromDb.getInitiator().getId().equals(userId)) { ...тут будет выброс исключения...`
Чтобы не дублировать код)

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,277 @@
+                "к событию не найдено событиес ID = % в БД.");
+        if (!eventFromDb.getEventState().equals(EventState.PUBLISHED)) {
+            //проверяем, что юзер - это инициатор события.
+            if (eventFromDb.getInitiator().getId().equals(eventId)) {
```
Здесь также можно было бы подправить условие, чтобы не дублировать код)

TyutterinYakov reviewed (commented):
Добрый вечер! Отметил оставшиеся недочеты) Хорошего вечера! Если возникнут вопросы - пиши в пачку, буду рад помочь🌃

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• после 2-го ревью.

laserdi commented on ewm-service/src/main/java/ru/practicum/explore_with_me/controller/CommentUserController.java:
```diff
@@ -0,0 +1,88 @@
+
+    @PostMapping("/user/{userId}")
+    @ResponseStatus(HttpStatus.CREATED)
+    public CommentForView addCommentByUser(@PathVariable("userId") @Positive @NotNull Long userId,
```
Убрал.

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• после 2-го ревью.
• подправил поиск комментариев для списка событий.

laserdi committed:
Diplom часть 3. Дополнительный сервис Комментарии.
• после 2-го ревью.
• подправил поиск комментариев для списка событий.

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/dto/comment/CommentEventRequest.java:
```diff
@@ -0,0 +1,16 @@
+@RequiredArgsConstructor
+@AllArgsConstructor
+public class CommentEventRequest {
+    List<Long> ids;
```
Для соблюдения инкапсуляции - поля класса должны быть приватными)

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,284 @@
+    public void deleteCommentByAdmin(Long comId) {
+        commentRepository.findById(comId).orElseThrow(() -> new NotFoundRecordInBD(String.format(
+                "При удалении комментария с ID = %d администратором комментарий не найден в БД.", comId)));
+        System.out.println("Найден комментарий перед удалением.");
```
Для логгирования мы теперь используем логгер)

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,284 @@
+        //Проверяем опубликовано ли событие
+        if (!eventFromDb.getEventState().equals(EventState.PUBLISHED)) {
+            //Событие не опубликовано.
+            if (eventFromDb.getInitiator().getId().equals(userId)) {
```
Аналогично, здесь также можно было бы изменить условие:
`if (!eventFromDb.getInitiator().getId().equals(userId)) { ...тут будет выброс исключения...`
Чтобы не дублировать код)

TyutterinYakov commented on ewm-service/src/main/java/ru/practicum/explore_with_me/service/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,284 @@
+                "При создании комментария пользователем в базе данных не найдено событие с ID = %d.");
+        //Проверяем опубликовано ли событие
+        if (!eventFromDb.getEventState().equals(EventState.PUBLISHED)) {
+            if (eventFromDb.getInitiator().getId().equals(userId)) {
```
Можно было подправить условие таким образом:
```
if (!eventFromDb.getInitiator().getId().equals(userId)) {
    log.info("Пользователь пытается сделать комментарий неопубликованного события.");
    throw new OperationFailedException("Пользователь пытается сделать комментарий " +
                            "неопубликованного события.");
}
```
Чтобы не дублировать код) Так как после условий идет такая же логика по сохранению комментария)

TyutterinYakov reviewed (approved):
Добрый вечер! Ну что ж, поздравляю с успешной сдачей финального спринта дипломного проекта! Ты большой молодец) Я рад, что ты справился со всеми встретившимися трудностями и дошел до конца! Желаю успехов в дальнейшем обучении и во всех начинаниях) Я отметил еще парочку моментов, что не были учтены, но думаю, что ты справишься с ними вне ревью) Обязательно их подправь, чтобы код проекта выглядел еще лучше) Прекрасного вечера и отличного настроения! Успехов на защите!🥳🎇🌁
